### PR TITLE
Fix audit v2 merge conflicts with master and rework audit_log API

### PR DESCRIPTION
**Pull request by noahmoss**
**Problem**

There were a great deal of merge conflicts with the `auditv2-main` feature branch, primarily because of [this PR](https://github.com/metabase/metabase/pull/34947). This PR was mainly working on the revision subsystem, not audit, but it changed the `publish-event!` API to expect a map of this structure:
```
{:object object, :user-id user-id, :other-keys ...}
```
rather than just an instance of a model itself (like a card, dashboard, etc). It also enforced this structure with new malli schemas for all the existing event types.

I like this approach, but it required adapting a lot of our `audit_log` code to use this new structure. In addition, since that PR was pretty large, it conflicted with audit v2 changes in other assorted ways which broke a number of tests (over 400 backend tests were failing once I finished the initial rebase).

**Approach to fixing this**

1. I've created a new branch `auditv2-main-2` which has a PR open here: https://github.com/metabase/metabase/pull/35434 This branch is just `auditv2-main` rebased on master, with a best-effort attempt to fix code conflicts. This doesn't do any larger architecture changes or test fixes, for the most part.
2. I've opened this PR with my changes to the way audit is structured, targeting that branch. This way, these changes can be reviewed in a slightly more visible way, instead of them just being hidden by a giant rebase. Once this PR is green I'll merge it into `auditv2-main-2` and we can use that as the authoritative feature branch.

**New things in this PR**
- `publish-event!` calls have been changed to take `:object` and `:user-id` keys.
- I restructured the `audit-log/record-event!` API to make it much more flexible. Now it's just a two-argument function that takes a `params` map, and most of the event handlers in `metabase.events.audit-log` can pass their data straight through to it. See the docstring for that function for more details (I've left a comment highlighting where it is)
- Some misc test fixes and renamings. Will self-review to shine light on some interesting areas.

### CODE REVIEW

File: src/metabase/api/dashboard.clj [hunk 1]
```diff
@@ -379,49 +379,6 @@
     (validation/check-embedding-enabled)
     (api/check-superuser)))
 
-(api/defendpoint PUT "/:id"
```
Comment (noahmoss) [inline src/metabase/api/dashboard.clj:382]: This was moved by a PR on master to elsewhere in the file but got re-added during the rebase, so I'm removing it again here.

File: resources/migrations/000_migrations.yaml [hunk 2]
```diff
@@ -15905,6 +15905,40 @@ databaseChangeLog:
...
+            constraintName: fk_recent_views_ref_user_id
+            referencedTableName: core_user
+            baseColumnNames: user_id
+            referencedColumnNames: id
+            onDelete: CASCADE
+
+  - changeSet:
+      id: v48.00-048
+      author: noahmoss
+      comment: Add foreign key on recent_views with onDelete CASCADE
```
Comment (noahmoss) [inline resources/migrations/000_migrations.yaml:15928]: Did this to fix a failing test which couldn't delete a user because it had entries in `recent_views`. It's not strictly necessary in the app itself though since users are never deleted, just deactivated.

File: enterprise/backend/test/metabase_enterprise/sso/integrations/saml_test.clj [hunk 3]
```diff
@@ -472,7 +472,7 @@
                         :model_id (:id new-user)
                         :topic    :user-invited
                         :user_id  nil}
-                       (audit-log-test/event :user-invited (:id new-user))))))
+                       (audit-log-test/latest-event :user-invited (:id new-user))))))
```
Comment (noahmoss) [inline enterprise/backend/test/metabase_enterprise/sso/integrations/saml_test.clj:475]: Renamed this function because it was shadowing another `event` variable in a test (I forget where now)

File: src/metabase/events/audit_log.clj [hunk 4]
```diff
@@ -1,100 +1,93 @@
...
-   [metabase.api.common :as api]
    [metabase.events :as events]
    [metabase.models.audit-log :as audit-log]
+   [metabase.util :as u]
    [methodical.core :as methodical]
    [toucan2.core :as t2]))
 
 (derive ::event :metabase/event)
 
-(defn maybe-prepare-update-event-data
```
Comment (noahmoss) [inline src/metabase/events/audit_log.clj:13]: Moved this logic into `metabase.models.audit-log` to make it a bit more flexible: now the `record-event!` params map can take `object` and `previous-object`, and it'll automatically record the changes if there are any

File: src/metabase/events/schema.clj [hunk 5]
```diff
@@ -87,15 +87,11 @@
...
                       [:map {:closed true}
-                       [:object [:fn #(t2/instance-of? :model/Database %)]]])
-      with-user      (mc/schema
-                      [:merge default-schema
-                       [:map {:closed true}
-                        [:user-id  pos-int?]]])]
-
+                       [:object [:fn #(t2/instance-of? :model/Database %)]]
+                       [:previous-object {:optional true} [:fn #(t2/instance-of? :model/Database %)]]
+                       [:user-id pos-int?]])]
```
Comment (noahmoss) [inline src/metabase/events/schema.clj:92]: Tweaked this schema so that all DB events can take `user-id`

File: src/metabase/models/audit_log.clj [hunk 6]
```diff
@@ -31,51 +34,73 @@
...
-                 :topic    unqualified-topic
-                 :details  details
-                 :model    model-name
-                 :model_id model-id
-                 :user_id  user-id)
-     ;; TODO: temporarily double-writing to the `activity` table, delete this in Metabase v48
-     (when-not (#{:card-read :dashboard-read :table-read :card-query :setting-update} unqualified-topic)
+  `:object` and `:previous-object` both have `model-details` called on them to determine which fields should be audited,
+  then they are added to `:details` before the event is recorded. `:previous-object` is only included if any audited fields
+  were updated."
```
Comment (noahmoss) [inline src/metabase/models/audit_log.clj:66]: Here's the biggest change of the file. Reworked this function to have a cleaner interface, with a `params` map of entirely optional keys.

File: enterprise/backend/test/metabase_enterprise/audit_app/pages/dashboard_subscriptions_test.clj [hunk 7]
```diff
@@ -35,7 +35,7 @@
                                                                     :collection_id collection-id}
                           PulseChannel          {channel-id :id}   {:pulse_id       pulse-id
                                                                     :channel_type   "email"
-                                                                    :details        {:emails ["amazing@fake.com"]}
+                                                                    :details        {:emails ["amazing@metabase.com"]}
```
Comment (johnswanson) [inline enterprise/backend/test/metabase_enterprise/audit_app/pages/dashboard_subscriptions_test.clj:38]: Just curious, this change is unrelated right? And you're changing it because we want to use domains we control to prevent accidents (if e.g. these tests were run in a context where we could actually send out emails)?

File: test/metabase/query_processor/middleware/process_userland_query_test.clj [hunk 8]
```diff
@@ -124,35 +124,32 @@
...
+                                  (:dataset_query card)
+                                  :info {:card-id (:id card)
+                                         :executed-by (mt/user->id :rasta)}))
 
-        (is (true? (t2/exists? :model/ViewLog :model "card" :model_id (:id card) :user_id (mt/user->id :rasta)))))))
+      (is (true? (t2/exists? :model/ViewLog :model "card" :model_id (:id card) :user_id (mt/user->id :rasta))))))
 
   (testing "skip if context is dashboard or collection"
     (mt/with-temp [:model/Card card {:dataset_query (mt/mbql-query users)}]
-      (mt/with-model-cleanup [:model/ViewLog]
```
Comment (johnswanson) [inline test/metabase/query_processor/middleware/process_userland_query_test.clj:146]: Out of curiosity, why remove the `with-model-cleanup` here?

**Review by johnswanson — APPROVED** [review]
I like the API change! I had a couple questions but they're just to satisfy my curiosity.

File: src/metabase/events/audit_log.clj [hunk 9]
```diff
@@ -1,100 +1,93 @@
...
 (derive :event/dashboard-remove-cards ::dashboard-card-event)
 
-;; TODO (crumb)
 (methodical/defmethod events/publish-event! ::dashboard-card-event
-  [topic {:keys [dashcards id]}]
+  [topic {:keys [object dashcards user-id] :as _event}]
   ;; we expect that the object has just a dashboard :id at the top level
   ;; plus a `:dashcards` attribute which is a vector of the cards added/removed
-  (let [details (-> (t2/select-one [:model/Dashboard :description :name] :id id)
+  (let [details (-> (select-keys object [:description :name :id])
```
Comment (dpsutton) [inline src/metabase/events/audit_log.clj:55]: I just noticed we have an N+1 query situation here. (`t2/select` N times rather than something like

```clojure
(methodical/defmethod events/publish-event! ::dashboard-card-event
  [topic {:keys [object dashcards user-id] :as _event}]
  ;; we expect that the object has just a dashboard :id at the top level
  ;; plus a `:dashcards` attribute which is a vector of the cards added/removed
  (let [cards   (when (seq dashcards)
                  (t2/select-fn->fn :id #(select-keys % [:name :description])
                                    :model/Card
                                    :id [:in (map :card_id dashcards)]))
        details (-> (select-keys object [:description :name :id])
                    (assoc :dashcards (for [{:keys [id card_id]} dashcards]
                                        (-> (cards card_id)
                                            (assoc :id id)
                                            (assoc :card_id card_id)))))]
    (audit-log/record-event! topic
                               {:details  details
                                :user-id  user-id
                                :model    :model/Dashboard
                                :model-id (u/id object)})))
```

File: enterprise/backend/test/metabase_enterprise/audit_app/pages/dashboard_subscriptions_test.clj (hunk 7 above)
Comment (noahmoss) [reply enterprise/backend/test/metabase_enterprise/audit_app/pages/dashboard_subscriptions_test.clj:38]: Actually, this test was failing for me in the REPL because of the `subscription-allowed-domains` setting in `metabase-enterprise.advanced-config.models.pulse-channel`. It was failing because locally I had an enterprise token set with all of the features enabled, but in CI I guess this test is run without the `:email-allow-list` feature, so the `fake.com` domain doesn't error. 

I thought that `metabase.com` was the default value I think that's actually just set by a different test. A better fix would be to just clear all premium features for the duration of this test, which I actually did in some other tests in this PR as well. (i.e. `(premium-features-test/with-premium-features #{} ...)`) I'll change this to use that approach as well.

File: test/metabase/query_processor/middleware/process_userland_query_test.clj (hunk 8 above)
Comment (noahmoss) [reply test/metabase/query_processor/middleware/process_userland_query_test.clj:146]: I think in general, using `with-model-cleanup` for AuditLog and ViewLog doesn't make too much sense. Lots of other tests are writing to these tables implicitly by triggering events, so the tables will be filled with "fake" data anyway. So cleaning up these models just adds work to the tests without any benefit.

File: src/metabase/events/audit_log.clj (hunk 9 above)
Comment (noahmoss) [reply src/metabase/events/audit_log.clj:55]: Good call, will update

**Review by qwef — APPROVED** [review]

**Review by qwef — APPROVED** [review]

File: src/metabase/models/audit_log.clj [hunk 10]
```diff
@@ -31,51 +34,73 @@
...
-       :topic      topic
-       :object     object
-       :details-fn (fn [_] details)
-       :user-id    user-id)))))
+       {:topic    topic
+        :object   object
+        :details  details
+        :model    model
+        :model-id model-id
+        :user-id  user-id}))))
```
Comment (escherize) [inline src/metabase/models/audit_log.clj:103]: I'd like to propose separating the generation of the value on 98:103 from actually doing the side effect. 

So, instead of:

```clojure
(record-event! topic params)
```

we would call:

```clojure
(actually-record-the-event! 
  (generate-event topic params))
```

----

## Why? 

It makes it easier to see and test what the outputs for record-object! are. Yes, now we can use the database to see, but it would makes it easier to see them in the repl, too. The tests that check it could be run in parallel, and would be way faster, because they don't rely on the db.

It feels like such a nice spot to separate the imperative shell from the complicated pure logic-y bit.

wdyt?

**Review by escherize — COMMENTED** [review]
Looks kinda familiar in a way, but also looks good to me.

Question: There are other handlers for `publish-event!`, do you know if changing its API will make those break? I assume if the tests are passing it should be OK. But I wanted to bring it up.

File: src/metabase/models/audit_log.clj (hunk 10 above)
Comment (escherize) [reply src/metabase/models/audit_log.clj:103]: In fact we could have a long test using e.g. `are` with 50 inputs/outputs and a simple `=` check for `generate-event` that is simple, and easy to follow (v.s. doing database setup/teardown/etc. for each one).

**Review by escherize — APPROVED** [review]
